src/ir/possible-contents.h

```diff
@@ -0,0 +1,577 @@
+    value = other.value;
+    return *this;
+  }
+
+  bool operator==(const PossibleContents& other) const {
+    return value == other.value;
+  }
+
+  bool operator!=(const PossibleContents& other) const {
+    return !(*this == other);
+  }
+
+  // Combine the information in a given PossibleContents to this one. The
+  // contents here will then include whatever content was possible in |other|.
+  void combine(const PossibleContents& other) {
+    // First handle the trivial cases of them being equal, or one of them is
+    // None or Many.
+    if (*this == other) {
+      return;
+    }
+    if (other.isNone()) {
+      return;
+    }
+    if (isNone()) {
+      value = other.value;
+      return;
+    }
+    if (isMany()) {
+      return;
+    }
+    if (other.isMany()) {
+      value = Many();
+      return;
+    }
+
+    auto type = getType();
+    auto otherType = other.getType();
+
+    if (!type.isRef() || !otherType.isRef()) {
+      // At least one is not a reference. We could in principle try to find
+      // ExactType here, say as the combination of two different constants, but
+      // as subtyping does not exist, Many is good enough anyhow, so do that.
+      value = Many();
+      return;
+    }
+
+    // Special handling for references from here.
+
+    // Nulls are always equal to each other, even if their types differ.
+    if (isNull() || other.isNull()) {
+      // If only one is a null, but the other's type is known exactly, then the
+      // combination is to add nullability (if the type is *not* known exactly,
+      // like for a global, then we cannot do anything useful here).
+      if (!isNull() && isTypeExact()) {
+        value = ExactType(Type(type.getHeapType(), Nullable));
+        return;
+      } else if (!other.isNull() && other.isTypeExact()) {
+        value = ExactType(Type(otherType.getHeapType(), Nullable));
+        return;
+      } else if (isNull() && other.isNull()) {
+        // Both are null. The result is a null, of the LUB.
+        auto lub = HeapType::getLeastUpperBound(type.getHeapType(),
+                                                otherType.getHeapType());
+        value = Literal::makeNull(lub);
+        return;
+      }
+    }
+
+    if (isTypeExact() && other.isTypeExact() &&
+        type.getHeapType() == otherType.getHeapType()) {
+      // We know the types here exactly, and even the heap types match, but
+      // there is some other difference that prevents them from being 100%
+      // identical (for example, one might be an ExactType and the other a
+      // Literal; or both might be ExactTypes and only one might be nullable).
+      // In these cases we can emit a proper ExactType here, adding nullability
+      // if we need to.
+      value = ExactType(Type(
+        type.getHeapType(),
+        type.isNullable() || otherType.isNullable() ? Nullable : NonNullable));
+      return;
+    }
+
+    // Nothing else possible combines in an interesting way; emit a Many.
```

Comment:
What about a literal and an exact type? Those combine to another exact type if the literal is of the same type.

edit: Ah, I see that this is covered since `isTypeExact` is different from `isExactType`.